server/tests/src/main/scala/sttp/tapir/server/tests/ServerMetricsTest.scala

```diff
@@ -0,0 +1,122 @@
+package sttp.tapir.server.tests
+
+import cats.effect.IO
+import cats.implicits._
+import org.scalatest.Assertion
+import org.scalatest.matchers.should.Matchers._
+import sttp.client3.{SttpBackend, _}
+import sttp.monad.MonadError
+import sttp.monad.syntax._
+import sttp.tapir.metrics.{EndpointMetric, Metric}
+import sttp.tapir.server.interceptor.metrics.MetricsRequestInterceptor
+import sttp.tapir.server.tests.ServerMetricsTest._
+import sttp.tapir.tests.TestUtil.inputStreamToByteArray
+import sttp.tapir.tests.{Test, _}
+
+import java.io.{ByteArrayInputStream, InputStream}
+
+class ServerMetricsTest[F[_], ROUTE, B](
+    backend: SttpBackend[IO, Any],
+    createServerTest: CreateServerTest[F, Any, ROUTE, B]
+)(implicit m: MonadError[F]) {
+  import createServerTest._
+
+  def tests(): List[Test] = List(
+    {
+      val reqCounter = newRequestCounter[F]
+      val resCounter = newResponseCounter[F]
+      val metrics = new MetricsRequestInterceptor[F, B](List(reqCounter, resCounter), Seq.empty)
+
+      testServer(in_json_out_json.name("metrics"), metricsInterceptor = metrics.some)(f =>
+        (if (f.fruit == "apple") Right(f) else Left(())).unit
+      ) { baseUri =>
+        basicRequest // onDecodeSuccess path
+          .post(uri"$baseUri/api/echo")
+          .body("""{"fruit":"apple","amount":1}""")
+          .send(backend)
+          .map { r =>
+            r.body shouldBe Right("""{"fruit":"apple","amount":1}""")
+            reqCounter.metric.value shouldBe 1
+            resCounter.metric.value shouldBe 1
+          } >> basicRequest // onDecodeFailure path
+          .post(uri"$baseUri/api/echo")
+          .body("""{"invalid":"body",}""")
+          .send(backend)
+          .map { _ =>
+            wait(200) {
```

Comment:
why do we have to wait for the counters to be updated?

Reply:
Without that it sometimes fail for akka server, maybe `transformDataBytes` where the callback is executed is called after response is returned, but it shouldn't right?

Reply:
Ah, onComplete can run asynchronously, after the request completes and is sent. Hm... maybe we can use `eventually`here? I wouldn't want to always sleep 200ms, this can be a cause of flaky tests